src/actors/common/cache.rs

```diff
@@ -167,138 +172,140 @@ impl<T: CacheItemRequest> Cacher<T> {
-                .clone()
-                .map_err(|_| {
-                    panic!("Oneshot channel cancelled! Race condition or system shutting down")
-                })
-                .and_then(|result| (*result).clone());
-
-            return Box::new(future);
+    /// Compute an item.
+    ///
+    /// If the item is in the file system cache, it is returned immediately. Otherwise, it is
+    /// computed using `T::compute`, and then persisted to the cache.
+    fn compute(&self, request: T, key: CacheKey) -> ResponseFuture<T::Item, T::Error> {
+        let cache_path = get_scope_path(self.config.cache_dir(), &key.scope, &key.cache_key);
+        if let Some(ref path) = cache_path {
+            if let Some(item) = tryf!(self.lookup_cache(&request, &key, &path)) {
+                return Box::new(future::ok(item));
+            }
         }
 
-        // A concurrent cache lookup is considered new. This does not imply a full cache miss.
-        metric!(counter(&format!("caches.{}.channel.miss", name)) += 1);
+        let name = self.config.name();
+        let key = request.get_cache_key();
+
+        // A file was not found. If this spikes, it's possible that the filesystem cache
+        // just got pruned.
+        metric!(counter(&format!("caches.{}.file.miss", name)) += 1);
 
-        let (tx, rx) = oneshot::channel();
+        let temp_file = if let Some(ref path) = cache_path {
+            let dir = path.parent().unwrap();
+            tryf!(fs::create_dir_all(dir));
+            tryf!(NamedTempFile::new_in(dir))
+        } else {
+            tryf!(NamedTempFile::new())
+        };
 
-        let slf = (*self).clone();
+        let future = request.compute(temp_file.path()).and_then(move |status| {
+            if let Some(ref cache_path) = cache_path {
+                sentry::configure_scope(|scope| {
+                    scope.set_extra(
+                        &format!("cache.{}.cache_path", name),
+                        cache_path.to_string_lossy().into(),
+                    );
+                });
 
-        let compute_future = lazy(clone!(key, || {
-            if let Some(item) = tryf!(slf.lookup_cache(&request)) {
-                return Box::new(Ok(item).into_future());
+                log::trace!("Creating {} at path {:?}", name, cache_path);
             }
 
-            // A file was not found. If this spikes, it's possible that the filesystem cache
-            // just got pruned.
-            metric!(counter(&format!("caches.{}.file.miss", name)) += 1);
+            let byteview = ByteView::open(temp_file.path())?;
 
-            let cache_path = get_scope_path(slf.config.cache_dir(), &key.scope, &key.cache_key);
+            metric!(
+                counter(&format!("caches.{}.file.write", name)) += 1,
+                "status" => status.as_ref(),
+            );
+            metric!(
+                time_raw(&format!("caches.{}.file.size", name)) = byteview.len() as u64,
```

Comment:
So there's a comment in the macro saying that `time_raw` is intentional for file sizes.  But why?

Reply:
We essentially use timer metrics across the board for measuring distributions of individual things. Potentially it would be better to use histograms.